### PR TITLE
Allow to re-trigger ChainSel 

### PR DESCRIPTION
**Pull request by Niols**

### CODE REVIEW

File: ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/Types.hs [hunk 1]
```diff
@@ -443,10 +445,10 @@ type FutureBlocks m blk = Map (HeaderHash blk) (Header blk, InvalidBlockPunishme
 -- | FIFO queue used to add blocks asynchronously to the ChainDB. Blocks are
 -- read from this queue by a background thread, which processes the blocks
 -- synchronously.
-newtype BlocksToAdd m blk = BlocksToAdd (TBQueue m (BlockToAdd m blk))
-  deriving NoThunks via OnlyCheckWhnfNamed "BlocksToAdd" (BlocksToAdd m blk)
+newtype ChainSelQueue m blk = ChainSelQueue (TBQueue m (ChainSelMessage m blk))
```
Comment (tek) [inline ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/Types.hs:448]: @amesgen pls review if this change is acceptable

Comment (amesgen) [reply ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/Types.hs:448]: Yes, this looks good :+1:

File: ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/Types.hs [hunk 2]
```diff
@@ -457,20 +461,27 @@ data BlockToAdd m blk = BlockToAdd
     -- ^ Used for the 'blockProcessed' field of 'AddBlockPromise'.
   }
 
--- | Create a new 'BlocksToAdd' with the given size.
-newBlocksToAdd :: IOLike m => Word -> m (BlocksToAdd m blk)
-newBlocksToAdd queueSize = BlocksToAdd <$>
+-- | Different async tasks for triggering ChainSel
+data ChainSelMessage m blk
+  -- | Add a new block
+  = ChainSelAddBlock (BlockToAdd m blk)
```
Comment (amesgen) [inline ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/Types.hs:467]:
```suggestion
  = ChainSelAddBlock !(BlockToAdd m blk)
```

File: ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs [hunk 3]
```diff
@@ -538,6 +555,14 @@ chainSelectionForBlock cdb@CDB{..} blockCache hdr punish = electric $ do
         traceWith addBlockTracer (TryAddToCurrentChain p)
         addToCurrentChain succsOf' curChainAndLedger maxExtra
 
+      -- The block @b@ fits onto the end of our current chain, but we can't add
+      -- it because the LoE limit was exceeded, so we store it in @cdbLoEBlocks@
+      -- for later reprocessing.
+      | pointHash tipPoint == headerPrevHash hdr -> do
+        atomically (modifyTVar cdbLoEBlocks (hdr :))
+        traceWith addBlockTracer (PostponedByLoE p)
+        return tipPoint
```
Comment (amesgen) [inline ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs:564]: This currently does not handle cases where blocks arrive out of order, ie we select `hdr`, but not some of its successors already in the VolDB. Idea: use `VolatileDB.filterByPredecessor`/`succsOf` to find all successors of the new tip, and add them to `cdbLoEBlocks`.

File: ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/Types.hs [hunk 4]
```diff
@@ -274,6 +277,7 @@ data ChainDbEnv m blk = CDB
     -- containing it. The fragment is usually anchored at a recent immutable
     -- tip; if it does not, it will conservatively be treated as the empty
     -- fragment anchored in the current immutable tip.
+  , cdbLoEBlocks       :: !(StrictTVar m [Header blk])
```
Comment (amesgen) [inline ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/Types.hs:280]: We should document what this is and some properties (it is empty if the LoE is disabled, blocks are only removed by GC, ...).

File: ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/Paths.hs [hunk 5]
```diff
@@ -293,8 +301,8 @@ data ReversePath blk =
 
 -- | Lazily compute the 'ReversePath' that starts (i.e., ends) with the given
 -- 'HeaderHash'.
-computeReversePath ::
-     forall blk.
+computeReversePath
+  :: forall blk.
```
Comment (amesgen) [inline ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/Paths.hs:305]: Boring style remark

File: ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/Paths.hs [hunk 6]
```diff
@@ -372,8 +380,8 @@ computeReversePath lookupBlockInfo endHash =
 --
 -- When the suffix of the 'ChainDiff' is non-empty, @P@ will be the last point
 -- in the suffix.
-isReachable ::
-     forall blk. (HasHeader blk, GetHeader blk)
+isReachable
+  :: forall blk. (HasHeader blk, GetHeader blk)
```
Comment (amesgen) [inline ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/Paths.hs:384]: Boring style remark

File: ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/Types.hs [hunk 7]
```diff
@@ -268,14 +268,12 @@ data ChainDbEnv m blk = CDB
     -- The number of blocks from the future is bounded by the number of
     -- upstream peers multiplied by the max clock skew divided by the slot
     -- length.
-  , cdbLoEFrag         :: !(StrictTVar m (AnchoredFragment (Header blk)))
+  , cdbLoE             :: LoE m blk
     -- ^ Fragment whose tip indicates the Limit on Eagerness, i.e. we are not
     -- allowed to select a chain from which we could not switch back to a chain
     -- containing it. The fragment is usually anchored at a recent immutable
     -- tip; if it does not, it will conservatively be treated as the empty
     -- fragment anchored in the current immutable tip.
```
Comment (amesgen) [inline ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/Types.hs:279]: The comment is now outdated, right?

File: ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs (hunk 3 above)
Comment (tek) [reply ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs:565]: ✅

File: ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/Types.hs (hunk 4 above)
Comment (tek) [reply ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/Types.hs:281]: ✅

File: ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/Paths.hs (hunk 5 above)
Comment (tek) [reply ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/Paths.hs]: ✅

File: ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/Types.hs (hunk 7 above)
Comment (tek) [reply ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/Types.hs]: ✅

File: ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs (hunk 3 above)
Comment (amesgen) [reply ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs:568]: a026aaafb2db062bd763fb48e1f41e851ca9c75c

Comment (tek) [reply ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/ChainSel.hs:568]: ✅

**Comment by Niols** [comment]
Kontinuierliche Integration ist grün :tada:

**Pull request merged**